packages/lesswrong/server/callbacks/rateLimits.ts

```diff
@@ -1,29 +1,187 @@
 import { Posts } from '../../lib/collections/posts'
 import { userIsAdmin, userIsMemberOf } from '../../lib/vulcan-users/permissions';
-import { DatabasePublicSetting } from '../../lib/publicSettings';
 import { getCollectionHooks } from '../mutationCallbacks';
 import Comments from '../../lib/collections/comments/collection';
 import { RATE_LIMIT_THREE_COMMENTS_PER_POST_PER_WEEK } from '../../lib/collections/moderatorActions/schema';
 import { getModeratorRateLimit, getTimeframeForRateLimit, userHasActiveModeratorActionOfType } from '../../lib/collections/moderatorActions/helpers';
 import moment from 'moment';
 import Users from '../../lib/collections/users/collection';
 import { captureEvent } from '../../lib/analyticsEvents';
+import { ForumOptions, forumSelect } from '../../lib/forumTypeUtils';
 
+type IntervalUnitType = 'seconds'|'minutes'|'hours'|'days'|'weeks'|'months'
+export type RateLimitType = "moderator"|"lowKarma"|"universal"|"downvoteRatio"
+ 
+export type RateLimitInfo = {
+  nextEligible: Date,
+  rateLimitType: RateLimitType,
+  rateLimitMessage: string,
+}
 
-const postIntervalSetting = new DatabasePublicSetting<number>('forum.postInterval', 30) // How long users should wait between each posts, in seconds
-const maxPostsPer24HoursSetting = new DatabasePublicSetting<number>('forum.maxPostsPerDay', 5) // Maximum number of posts a user can create in a day
+/* 
+Each forum can set a list of automatically applied rate limits. 
+
+Whenever a user submits a post or comment, the server checks if any of the listed AutoRateLimits 
+apply to that user/post/comment.
+
+AutoRateLimits check how documents the user has posted in a recent time interval, and prevents them
+from posting more if they've posted more than the alloted number of itemsPerInterval.
+
+AutoRateLimits can take in an optional karmaThreshold or downvoteRatio parameter. If set, the AutoRateLimit
+applies to users who meet that karmaThreshold and/or downvoteRatio criteria. If both params are set, the 
+rate limit only applies if both conditions are met. If neither param is set, the rate limit applies to all users.
+*/
+interface AutoRateLimit {
+  karmaThreshold?: number, // if set, the rate limit will only apply to users with karma less than the threshold
+  downvoteRatio?: number, // if set, the rate limit will only apply to users who's ratio of received downvotes / total votes is higher than the listed threshold
+  timeframeLength: number, // how long the time timeframe is (measured in the intervalUnit, below)
+  timeframeUnit: IntervalUnitType, // measuring units for the timeframe (i.e. minutes, hours, days)
+  itemsPerTimeframe: number, // number of items a user can post/comment/etc before triggering rate limit
+  rateLimitType: RateLimitType // short name used in analytics db
+  rateLimitMessage: string // A message displayed to users when they are rate limited
+}
+
+type ForumAutoRateLimits = {
```

Comment:
I think I'd probably put the content/action type into `AutoRateLimit` as another field to be configured, rather than having two top-level post & comment arrays - that's how the new `UserRateLimits` table is structured.  When I first scanned through `autoRateLimits` I was confused and thought it was missing entirely.